### PR TITLE
Upgrade `Subnet` API version from `2022-07-01` to `2023-09-01` and switch to go-azure-sdk

### PR DESCRIPTION
**Pull request by neil-yechenwei**
Service team is reaching us for supporting new enum values "NetworkSecurityGroupEnabled" and "RouteTableEnabled" for the "[privateEndpointNetworkPolicies](https://github.com/Azure/azure-rest-api-specs/blob/77df3b1680a95566a3e6a3a2cd07c9a4960b2af3/specification/network/resource-manager/Microsoft.Network/stable/2023-09-01/virtualNetwork.json#L1602)" property in "azurerm_subnet". This new feature only exists in the latest API version 2023-09-01. So I changed SubnetsClient to use 2023-09-01 since currently it is referring 2022-07-01.

Related reference: https://github.com/hashicorp/terraform-provider-azurerm/pull/25064

This PR impacts below RPs.
web:
image

network:
image

### CODE REVIEW

**Comment by katbyte** [comment]
and a bunch nmore
```
------- Stdout: -------
=== RUN   TestAccVirtualNetworkGateway_vpnGw2
=== PAUSE TestAccVirtualNetworkGateway_vpnGw2
=== CONT  TestAccVirtualNetworkGateway_vpnGw2
    testcase.go:113: Step 1/1 error: Error running apply: exit status 1
        
        Error: Creating/Updating Virtual Network Gateway: (Name "acctestvng-240408165942485177" / Resource Group "acctestRG-240408165942485177"): network.VirtualNetworkGatewaysClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="PublicIpWithBasicSkuNotAllowedOnVPNGateways" Message="Basic IP configuration for VPN Virtual Network Gateways is not supported. Follow the link for more details : https://go.microsoft.com/fwlink/p/?linkid=2241350 /subscriptions/*******/resourceGroups/acctestRG-240408165942485177/providers/Microsoft.Network/virtualNetworkGateways/acctestvng-240408165942485177" Details=[]
        
          with azurerm_virtual_network_gateway.test,
          on terraform_plugin_test.tf line 51, in resource "azurerm_virtual_network_gateway" "test":
          51: resource "azurerm_virtual_network_gateway" "test" {
        
--- FAIL: TestAccVirtualNetworkGateway_vpnGw2 (76.26s)
FAIL
```

**Review by katbyte — APPROVED** [review]
LGTM 🚋

**Comment by neil-yechenwei** [comment]
@katbyte , is this PR good to be merged?

File: internal/services/network/virtual_network_resource.go [hunk 1]
```diff
@@ -541,18 +541,26 @@ func resourceAzureSubnetHash(v interface{}) int {
...
-	return &resp, nil
+	for _, subnet := range pointer.From(resp.Subnets) {
+		if subnetName == pointer.From(subnet.Name) {
+			// Return it directly rather than copy the fields to prevent potential uncovered properties (for example, `ServiceEndpoints` mentioned in #1619)
+			return pointer.To(subnet), nil
+		}
+	}
+
+	// Return empty object when the Subnet isn't found
+	return pointer.To(network.Subnet{}), nil
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply internal/services/network/virtual_network_resource.go]: Updated

File: internal/services/network/subnet_resource.go [hunk 2]
```diff
@@ -649,26 +646,26 @@ func resourceSubnetDelete(d *pluginsdk.ResourceData, meta interface{}) error {
...
-	future, err := client.Delete(ctx, id.ResourceGroupName, id.VirtualNetworkName, id.SubnetName)
+	future, err := client.Delete(ctx, *id)
 	if err != nil {
 		return fmt.Errorf("deleting %s: %+v", *id, err)
 	}
 
-	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
-		if !response.WasNotFound(future.Response()) {
+	if err = future.Poller.PollUntilDone(ctx); err != nil {
+		if !response.WasNotFound(future.HttpResponse) {
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply internal/services/network/subnet_resource.go]: Updated to use DeleteThenPoll.

**Comment by neil-yechenwei** [comment]
@stephybun , thanks for the comments. I updated PR and just now re-triggered the related test cases. Below is the latest test result. Please take another look. Thanks.

image
image
image

**Review by stephybun — APPROVED** [review]
Thanks @neil-yechenwei LGTM 👍

**Pull request merged**